Source/WebGPU/WebGPU/Sampler.mm

```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
     return !!m_samplerIdentifier;
 }
 
+NSUInteger Sampler::aliveCount()
+{
+    Locker locker { samplerStateLock };
+    [weakSamplerStates _markNeedsCompaction];
+    [weakSamplerStates compact];
+    return weakSamplerStates.count;
+}
+
 id<MTLSamplerState> Sampler::samplerState() const
 {
     if (!m_samplerIdentifier)
         return nil;
 
-    Locker locker { samplerStateLock };
-    if (!cachedSamplerStates) {
-        cachedSamplerStates = [NSMutableDictionary dictionary];
-        lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+    {
+        Locker locker { samplerStateLock };
+        if (!cachedSamplerStates) {
+            cachedSamplerStates = [NSMutableDictionary dictionary];
+            lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+            weakSamplerStates = [NSPointerArray weakObjectsPointerArray];
```

Comment:
Your commit messaqge does not really describe what problem you are trying to solve with weakSamplerStates.

It does not help that the testcase is 33k obfuscated blob.

Is the problem that even though WebGPU has freed its internal cache entries, some other part of the system might still hold the refs to the samplers in the freed cache entries, and thus freeing did nothing? And you solve this by tracking the weak ptrs?

Reply:
yes, effectively Metal command buffers will still retain references to the MTLSamplerStates in use by the command buffers, so a long running Metal command buffer can cause this race to occur with the existing sampler state cache.